rdflib/graph.py

```diff
@@ -2029,9 +2115,11 @@ def contexts(self, triple=None):
 
     def quads(self, quad):
         for s, p, o, c in super(Dataset, self).quads(quad):
-            if c.identifier == self.default_context:
+            if c.identifier == DATASET_DEFAULT_GRAPH_ID:
                 yield s, p, o, None
-            else:
+            elif len(quad) == 3:
+                yield s, p, o, c.identifier
```

Comment:
Sorry, I don't understand your point. All three implementations of `quads` are iterators.

Reply:
I'm saying that `ConjunctiveGraph.quads` returns a `Graph` for the fourth member of each quad, but `Dataset.quads` returns an `Identifier`. Admittedly, that isn't really the main point of this PR, so that may be a thing to consider later.

Reply:
Ah, you did explicitly write "return a Graph for the context", sorry for being obtuse. You're correct that the current [Dataset quads implementation returns an identifier](https://github.com/RDFLib/rdflib/blob/b2fdaf5a1f45c09694dbd8925ab6b6dee84436b4/rdflib/graph.py#L2030) but that behaviour [is documented](https://github.com/RDFLib/rdflib/blob/b2fdaf5a1f45c09694dbd8925ab6b6dee84436b4/rdflib/graph.py#L1884) which is sort of the _de facto_ definition of the public API.

I guess if this is to be changed, it needs to have a period of deprecation.

Reply:
mypy has just reminded me that if quads are going to have a Graph as a context, then the type signatures of `add` and `remove` need to reflect that - hence my original typing of `Optional[Union["Graph", "Identifier"]]`